trie/iterator.go

```diff
@@ -967,51 +964,125 @@ func (it *merkleTreeIterator) seek(path []byte) {
+		switch bytes.Compare(path, zk.NewTreePathFromHashBig(common.BytesToHash(leaf.key))) {
+		case -1: // path < leaf path
+		loop1:
+			for it.prevLeaf() {
+				switch bytes.Compare(path, zk.NewTreePathFromHashBig(common.BytesToHash(it.lastNodeAsLeaf().key))) {
+				case -1: // path < leaf path
+					continue
+				case 0: // path == leaf path
+					break loop1
+				case 1:
+					it.nextLeaf()
+					break loop1
+				}
+			}
+		case 0: // leaf path == path
+		case 1: // leaf path < path
+		loop2:
+			for it.nextLeaf() {
+				switch bytes.Compare(path, zk.NewTreePathFromHashBig(common.BytesToHash(it.lastNodeAsLeaf().key))) {
+				case -1: // path < leaf path
+					it.prevLeaf()
+					break loop2
+				case 0: // path == leaf path
+					break loop2
+				case 1:
+					continue
+				}
+			}
+			return
 		}
 	}
 
 	if len(it.path) == 0 {
-		return // root node is not parent node
+		return
 	}
-
 	// When using the Next function, it always moves to the next node.
 	// Therefore, it sets the previous node of the node retrieved by seek to the last visited node.
 	lastIdx := len(it.path) - 1
 	if it.path[lastIdx] == right {
 		it.path[lastIdx] = left
 		it.stack[len(it.stack)-1] = it.resolveNode(it.parentOfLastNode().children[left])
-		for {
-			if parent, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorParentNode); ok {
-				for _, path := range []byte{right, left} {
-					if child := it.resolveNode(parent.children[path]); child != nil {
-						it.stack = append(it.stack, child)
-						it.path = append(it.path, path)
-						break
-					}
-					if it.err != nil {
-						return
-					}
-				}
-			} else {
-				break
-			}
-		}
+		it.visitBeforeLeafNode()
 	} else {
 		it.path = it.path[:lastIdx]
 		it.stack = it.stack[:len(it.stack)-1]
 	}
 }
 
+func (it *merkleTreeIterator) visitBeforeLeafNode() {
+	for {
+		if parent, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorParentNode); ok {
+			for _, path := range []byte{right, left} {
+				if child := it.resolveNode(parent.children[path]); child != nil {
+					it.stack = append(it.stack, child)
+					it.path = append(it.path, path)
+					break
+				}
+				if it.err != nil {
+					return
+				}
+			}
+		} else {
+			break
+		}
+	}
+}
+
+func (it *merkleTreeIterator) prevLeaf() bool {
+	if _, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorLeafNode); ok {
+		for len(it.path) != 0 { // Infinite loop if there are still paths left to visit
+			switch lastPathIndex := len(it.path) - 1; it.path[lastPathIndex] {
+			case left: // left visited. go up
+				it.path = it.path[:len(it.path)-1]
+				it.stack = it.stack[:len(it.stack)-1]
+			case right: // right visited. go left
+				if leftNode := it.resolveNode(it.parentOfLastNode().children[left]); leftNode != nil {
+					it.path[lastPathIndex] = left
+					it.stack[len(it.stack)-1] = leftNode
+					it.visitBeforeLeafNode()
+					return true
+				}
+				if it.err != nil {
+					return false
```

Comment:
If an error occurred, `prevLeaf` returns false. Don't we need to add `if it.err != nil { return }` in the caller function `seek`?

Reply:
updated